File: prakticke/pu07/pu07-python/hamiltonianCycleTest.py
#! /usr/bin/env python3

import sys
import time
import traceback
import random
import itertools

from HamiltonianCycle import HamiltonianCycle

""" Nastavte na True ak chcete aby testy zastali na prvej chybe. """
stopOnError = False

class FailedTestException(BaseException):
    pass

def now():
    try:
       return time.perf_counter() # >=3.3
    except AttributeError:
       return time.time() # this is not monotonic!

def printException():
    print('ERROR: Exception raised:\n%s\n%s\n%s' % (
        '-'*20,
        traceback.format_exc(),
        '-'*20)
    )

class Tester(object):
    def __init__(self):
        self.case = 0
        self.tested = 0
        self.passed = 0
        self.time = 0

    def fail(self, msg):
        print('  ✕ Failed: %s' % msg)
        return False;

    def checkGood(self, m, r):
        if len(m) != len(r):
            return self.fail('Wrong result length!')
        last = r[-1]
        if last not in range(len(m)):
            return self.fail('Wrong vertex number in output: %s' % last)
        for v in r:
            if v not in range(len(m)):
                return self.fail('Wrong vertex number in output: %s' % v)
            if not m[last][v]:
                return self.fail('No edge %d -> %d' % (last, v))
            last = v
        if frozenset(range(len(m))) != frozenset(r):
            return self.fail("Missing some verticies");
        return True

    def checkBad(self, r):
        if len(r) != 0:
            return self.fail('non-empty result when there should be no cycle')
        return True

    def check(self, m, good, r):
        if good:
            return self.checkGood(m, r)
        else:
            return self.checkBad(r)


    def test(self, name, m, good):
        self.case += 1
        self.tested += 1
        print('Case %d (%s):  ' % (self.case, name))

        try:
            rs = random.getstate()
            start = now()
            r = HamiltonianCycle().find(m)
            duration = now() - start
            self.time += duration
            random.setstate(rs)
        except KeyboardInterrupt:
            raise KeyboardInterrupt()
        except:
            printException()
            if stopOnError:
                raise FailedTestException()
            return


        if self.check(m, good, r):
            self.passed += 1
            print('  ✓ PASSED  duration %1.3fms' % (duration*1000.0))
        else:
            print('    edges:')
            for row in m:
                print('    ' + ' '.join(('1' if x else '0') for x in row))
            print('')
            print('    result: %s' % repr(r))
            print('')
            if stopOnError:
                raise FailedTestException()

    def status(self):
        print()
        print("TESTED %d" % (self.tested,))
        print("PASSED %d" % (self.passed,))
        print("TIME %1.3fms" % (self.time * 1000.0))
        if self.tested == self.passed:
            print("OK")
            return True
        else:
            print("ERROR")
            return False


def edgesToIncidenceMatrix(edges):
    a, b = zip(*edges)
    n = max(a+b)+1

    m = []
    for row in range(n):
        m.append([False] * n)
    for a,b in edges:
        m[a][b] = True
    return m


def randomGoodInput(size):
    maxv = size-1
    m = edgesToIncidenceMatrix([(maxv, maxv)])
    for r in range(size):
        for c in range(size):
                m[r][c] = random.randint(0,1) == 1

    sol = list(range(size))
    random.shuffle(sol)
    last = sol[-1]
    for v in sol:
        m[last][v] = True
        last = v
    return m


def randomBadInput(size):
    def goodPath(m, path):
        last = path[-1]
        for v in path:
            if not m[last][v]:
                return False
            last = v
        return True

#    print('Generating BAD input of size %d' % size)
    m = []
    for r in range(size):
        m.append([True]*size)

    for perm in itertools.permutations(range(size)):
        if goodPath(m, perm):
            i = random.randrange(len(perm))
            j = (i+1)%len(perm)
            m[perm[i]][perm[j]] = False
#    print('Done generating')
#    print(repr(m))
    return m

random.seed(47)

t = Tester()
try:

    m = [
            [True, True],
            [True, True]
        ]
    t.test("K2", m, True)

    m = [
            [False, True],
            [False, False],
        ]
    t.test("2 -> 1", m, False)

    m = edgesToIncidenceMatrix([
            (0, 3),
            (0, 4),
            (1, 0),
            (1, 3),
            (1, 4),
            (2, 1),
            (2, 5),
            (3, 0),
            (3, 4),
            (4, 1),
            (4, 2),
            (4, 5),
            (5, 2),
            (5, 4),
        ])

    t.test("SomeGraph", m, True)

    herschel = [
            ( 1,  2),
            ( 1,  4),
            ( 1,  5),
            ( 2,  3),
            ( 2,  6),
            ( 2,  7),
            ( 3,  4),
            ( 3,  8),
            ( 4,  9),
            ( 4, 10),
            ( 5,  6),
            ( 5, 10),
            ( 6,  0),
            ( 7,  8),
            ( 7,  0),
            ( 8,  9),
            ( 9,  0),
            (10,  0),
        ]
    m = edgesToIncidenceMatrix(herschel + [ (b,a) for (a,b) in herschel ])

    t.test("Herschel graph", m, False)

    t.test("random good 10.1", randomGoodInput(10), True)
    t.test("random good 10.2", randomGoodInput(10), True)
    t.test("random good 10.3", randomGoodInput(10), True)
    t.test("random good 10.4", randomGoodInput(10), True)

    t.test("random good 20", randomGoodInput(20), True)

    t.test("random good 50", randomGoodInput(50), True)

    t.test("random bad 5.1", randomBadInput(5), False)
    t.test("random bad 5.2", randomBadInput(5), False)
    t.test("random bad 5.3", randomBadInput(5), False)
    t.test("random bad 5.4", randomBadInput(5), False)

    t.test("random bad 10", randomBadInput(10), False)

    m = [
        [True, True, True, True, True, False, False, True, True, True, True, False, True, False, False],
        [False, True, False, False, False, True, False, False, True, True, True, True, False, True, True],
        [False, False, True, False, False, False, False, False, True, False, True, False, True, False, False],
        [False, False, False, True, False, True, False, False, True, False, True, False, True, False, True],
        [False, False, False, False, True, True, False, False, False, False, True, True, True, True, True],
        [False, False, False, False, False, True, False, False, True, False, True, True, True, True, True],
        [False, False, False, False, False, False, True, False, True, True, True, True, True, True, True],
        [False, False, False, False, False, False, False, True, False, True, True, True, True, True, True],
        [False, False, False, False, False, False, False, False, True, False, True, False, False, True, True],
        [False, False, False, False, True, True, False, True, False, True, True, True, True, True, True],
        [False, False, True, True, True, True, False, False, False, True, True, True, True, True, False],
        [False, False, True, False, False, True, False, True, True, True, False, True, False, False, False],
        [False, False, True, True, True, True, False, True, True, True, False, False, True, False, False],
        [True, False, True, True, True, True, True, True, True, True, False, False, False, True, False],
        [False, True, True, True, False, True, False, False, True, False, False, False, False, False, True]
    ]
    t.test("BigFalse", m, False)

except FailedTestException:
    print("Stopped on first failed test!")
except:
    printException()
finally:
    sys.exit(0 if t.status() else 1)

# vim: set sw=4 ts=4 sts=4 et :

File: prakticke/pu07/pu07-java/HamiltonianCycleTest.java
import java.lang.Math;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Collections;
import java.util.Deque;
import java.util.function.Consumer;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Tester {
    int tested = 0;
    int passed = 0;
    int ncase = 0;
    long time = 0;
    Deque<Scope> scopes = new ArrayDeque<Scope>();
    Scope lastScope = null;

    class Scope implements AutoCloseable{
        private final Tester t;
        public final String name;
        public final Deque<String> messages = new ArrayDeque<String>();
        Scope(Tester t, String name) {
            this.t = t;
            this.name = name;
            t.scopes.push(this);
        }
        public void close() {
            t.scopes.pop();
            t.lastScope = t.scopes.peek();
        }
        public void message(String msg) {
            messages.push(msg);
        }
        public void print(String prefix) {
            if (!name.isEmpty())
                System.err.println(prefix + "In " + name);
            messages.descendingIterator().forEachRemaining(msg -> System.err.println(prefix + "  " + msg));
        }
    }

    public Scope scope(String msg) {
        lastScope = new Scope(this, msg);
        return lastScope;
    }
    public void message(String msg) {
        // throws if there isn't a scope
        lastScope.message(msg);
    }
    public void printScopes(String prefix) {
        scopes.descendingIterator().forEachRemaining(s -> s.print(prefix + "  "));
    }

    public boolean compare(Object result, Object expected, String msg) {
        tested++;
        if (expected.equals(result)) {
            passed++;
//            System.err.println("  ✓ " + msg);
            return true;
        } else {
            printScopes("  ");
            System.err.println("  ✕ Failed: " + msg + ":");
            System.err.println("      got " + result + " expected " + expected);
            return false;
        }
    }

    public <T> boolean compareRef(T result, T expected, String msg) {
        tested++;
        if (result == expected) {
            passed++;
//            System.err.println("  ✓ " + msg);
            return true;
        } else {
            printScopes("  ");
            System.err.println("  ✕ Failed: " + msg + ":");
            System.err.println("      got " + result + " expected " + expected);
            return false;
        }
    }

    public boolean testNotEqual(Object a, Object b, String msg) {
        return compare(!a.equals(b), true, msg);
    }

    public boolean testNotEqual(Object a, Object b) {
        return compare(!a.equals(b), true, a + " != " + b);
    }

    public boolean verify(boolean cond, String msg) {
        tested++;
        if (cond) {
            passed++;
            return true;
        } else {
            printScopes("  ");
            System.err.println("  ✕ Failed: " + msg);
            return false;
        }
    }

    public void fail(String msg) {
        tested++;
        printScopes("  ");
        System.err.println("  ✕ Failed: " + msg);
    }

    public void startCase(String s) {
        System.err.println(String.format("CASE %d: %s", ++ncase, s));
    }

    public void scope(String name, Consumer<Scope> c) {
        try (Tester.Scope s = scope(name)) {
            try {
                c.accept(s);
            }
            catch (Throwable e) {
                fail("Exception: " + e.toString());
                e.printStackTrace();
            }
        }
    }

    public void scope(Consumer<Scope> c) {
        scope("", c);
    }

    public void testCase(String name, Consumer<Scope> c) {
        startCase(name);
        scope(name, c);
    }

    public boolean checkGood(boolean[][] m, List<Integer> r) {
        int N = m.length;
        if (!compare(r.size(), N, "Wrong result length"))
            return false;
        int last = r.get(r.size() - 1);
        if (!verify(last < N, "Wrong vertex number in output: " + last))
            return false;
        for (int v : r) {
            if (!verify(v < N, "Wrong vertex number in output: " + v))
                return false;
            if (!verify(m[last][v], String.format("No edge %d -> %d", last, v)))
                return false;
            last = v;
        }
        Set<Integer> vs = new HashSet<>(r);
        for (int v = 0 ; v < N; ++v) {
            if (!verify(vs.contains(v), "Missing vertex " + v))
                return false;
        }
        return true;
    }

    public boolean checkBad(boolean[][] m, List<Integer> r) {
        return verify(r.size() == 0, "non-empty result when there should be no cycle");
    }
    public boolean check(boolean[][] m, boolean good, List<Integer> r) {
        tested++;
        if (good) {
            passed++;
            return checkGood(m, r);
        } else {
            return checkBad(m, r);
        }
    }

    public static String toString(boolean[] l) {
        return IntStream.range(0, l.length)
            .mapToObj(i -> l[i] ? "1" : "0")
            .collect(Collectors.joining(" "));
    }

    public static String toString(boolean[][] m) {
        return Arrays.stream(m)
            .map(l -> toString(l))
            .collect(Collectors.joining("\n        "))
        ;
    }

    public void test(String desc, boolean[][] m, boolean good) {
        testCase(desc, s -> {
            long start = System.nanoTime();
            List<Integer> r = (new HamiltonianCycle()).find(m);
            long duration = (System.nanoTime() - start) / 1000;
            time += duration;

            message("m: \n        " + toString(m) );
            message("result: " + r);
            if (check(m, good, r))
                System.err.println("  ✓ PASSED in " + (duration / 1000.0) + "ms");
        });
    }

    public boolean status() {
        System.err.println("");
        System.err.println("TESTED " + tested);
        System.err.println("PASSED " + passed);
        System.err.println("SUM(time) " + (time / 1000.0) + "ms");

        System.err.println(tested == passed ? "OK" : "ERROR" );
        return tested == passed;
    }

}


public class HamiltonianCycleTest {
    public static class Edge {
        int f, t;
        Edge(int f, int t) { this.f = f; this.t = t;}
    };
    public static Edge E(int f, int t) { return new Edge(f, t); }

    public static boolean[][] edgesToIncidenceMatrix(List<Edge> edges) {
        int max = 1 + edges.stream()
            .mapToInt(e -> Math.max(e.f, e.t))
            .max()
            .orElse(0)
        ;
        boolean[][] m = new boolean[max][max];
        for (Edge e : edges)
            m[e.f][e.t] = true;
        return m;
    }

    public static boolean[][] randomGoodInput(Random rng, int size) {
        boolean[][] m = new boolean[size][size];
        for (int r=0; r<size; ++r)
            for (int c=0; c<size; ++c)
                m[r][c] = rng.nextBoolean();

        List<Integer> path = IntStream.range(0, size).boxed().collect(Collectors.toList());
        Collections.shuffle(path, rng);
        int last = path.get(path.size() - 1);
        for (int v : path) {
            m[last][v] = true;
            last = v;
        }
        return m;
    }

    public static void main(String[] args) {
        Tester t = new Tester();
        Random rng = new Random(47);
        {
            boolean[][] m = {{true, true}, {true, true}};
            t.test("K2", m, true);
        }
        {
            boolean[][] m = {{false, true}, {false, false}};
            t.test("2->1", m, false);
        }
        {
            boolean[][] m = edgesToIncidenceMatrix(Arrays.asList(
                    E(0, 3),
                    E(0, 4),
                    E(1, 0),
                    E(1, 3),
                    E(1, 4),
                    E(2, 1),
                    E(2, 5),
                    E(3, 0),
                    E(3, 4),
                    E(4, 1),
                    E(4, 2),
                    E(4, 5),
                    E(5, 2),
                    E(5, 4)
                ));

            t.test("SomeGraph", m, true);
        }
        {
            boolean[][] m = edgesToIncidenceMatrix(Arrays.asList(
                    E( 1,  2), E( 2,  1),
                    E( 1,  4), E( 4,  1),
                    E( 1,  5), E( 5,  1),
                    E( 2,  3), E( 3,  2),
                    E( 2,  6), E( 6,  2),
                    E( 2,  7), E( 7,  2),
                    E( 3,  4), E( 4,  3),
                    E( 3,  8), E( 8,  3),
                    E( 4,  9), E( 9,  4),
                    E( 4, 10), E(10,  4),
                    E( 5,  6), E( 6,  5),
                    E( 5, 10), E(10,  5),
                    E( 6,  0), E( 0,  6),
                    E( 7,  8), E( 8,  7),
                    E( 7,  0), E( 0,  7),
                    E( 8,  9), E( 9,  8),
                    E( 9,  0), E( 0,  9),
                    E(10,  0), E( 0, 10)
                ));

            t.test("Herschel graph", m, false);
        }

        t.test("random good 10.1", randomGoodInput(rng, 10), true);
        t.test("random good 10.2", randomGoodInput(rng, 10), true);
        t.test("random good 10.3", randomGoodInput(rng, 10), true);
        t.test("random good 10.4", randomGoodInput(rng, 10), true);

        t.test("random good 20", randomGoodInput(rng, 20), true);

        t.test("random good 50", randomGoodInput(rng, 50), true);

        {
            boolean[][] m = {
                {true, true, true, true, true, false, false, true, true, true, true, false, true, false, false},
                {false, true, false, false, false, true, false, false, true, true, true, true, false, true, true},
                {false, false, true, false, false, false, false, false, true, false, true, false, true, false, false},
                {false, false, false, true, false, true, false, false, true, false, true, false, true, false, true},
                {false, false, false, false, true, true, false, false, false, false, true, true, true, true, true},
                {false, false, false, false, false, true, false, false, true, false, true, true, true, true, true},
                {false, false, false, false, false, false, true, false, true, true, true, true, true, true, true},
                {false, false, false, false, false, false, false, true, false, true, true, true, true, true, true},
                {false, false, false, false, false, false, false, false, true, false, true, false, false, true, true},
                {false, false, false, false, true, true, false, true, false, true, true, true, true, true, true},
                {false, false, true, true, true, true, false, false, false, true, true, true, true, true, false},
                {false, false, true, false, false, true, false, true, true, true, false, true, false, false, false},
                {false, false, true, true, true, true, false, true, true, true, false, false, true, false, false},
                {true, false, true, true, true, true, true, true, true, true, false, false, false, true, false},
                {false, true, true, true, false, true, false, false, true, false, false, false, false, false, true}
            };
            t.test("BigFalse", m, false);
        }

        System.exit(t.status() ? 0 : 1);
    }
}

